src/analysis/complex/upper_half_plane.lean

```diff
@@ -0,0 +1,165 @@
+
+/-!
+# The upper half plane and its automorphisms
+
+This file defines `upper_half_plane` to be the upper half plane in `ℂ`.
+
+We furthermore equip it with the structure of an `SL(2,ℝ)` action by
+fractional linear transformations.
+
+We define the notation `ℍ` for the upper half plane available in the locale
+`upper_half_plane` so as not to conflict with the quaternions.
+-/
+
+noncomputable theory
+
+open matrix matrix.special_linear_group
+
+open_locale classical big_operators matrix_groups
+
+local attribute [instance] fintype.card_fin_even
+
+/-- The upper half plane -/
+abbreviation upper_half_plane :=
+{point : ℂ // 0 < point.im}
+
+localized "notation `ℍ` := upper_half_plane" in upper_half_plane
+
+namespace upper_half_plane
+
+/-- Imaginary part -/
+def im (z : ℍ) := (z : ℂ).im
+
+/-- Real part -/
+def re (z : ℍ) := (z : ℂ).re
+
+@[simp] lemma coe_im (z : ℍ) : (z : ℂ).im = z.im := rfl
+
+@[simp] lemma coe_re (z : ℍ) : (z : ℂ).re = z.re := rfl
+
+lemma im_pos (z : ℍ) : 0 < z.im := z.2
+
+lemma im_ne_zero (z : ℍ) : z.im ≠ 0 := z.im_pos.ne'
+
+lemma ne_zero (z : ℍ) : (z : ℂ) ≠ 0 :=
+begin
+  intros h,
+  apply z.im_ne_zero,
+  simp [← complex.zero_im, ← h]
+end
+
+lemma norm_sq_pos (z : ℍ) : 0 < complex.norm_sq (z : ℂ) :=
+begin
+  rw complex.norm_sq_pos,
+  exact z.ne_zero
+end
+
+lemma norm_sq_ne_zero (z : ℍ) : complex.norm_sq (z : ℂ) ≠ 0 := ne_of_gt (norm_sq_pos z)
+
+/-- Numerator of the formula for a fractional linear transformation -/
+@[simp] def top (g : SL(2, ℝ)) (z : ℍ) : ℂ := (g 0 0) * z + (g 0 1)
+
+/-- Denominator of the formula for a fractional linear transformation -/
+@[simp] def bottom (g : SL(2, ℝ)) (z : ℍ) : ℂ := (g 1 0) * z + (g 1 1)
+
+lemma linear_ne_zero (cd : fin 2 → ℝ) (z : ℍ) (hcd : cd ≠ 0) : (cd 0 : ℂ) * z + cd 1 ≠ 0 :=
+begin
+  intros h,
+  apply hcd,
+  have : cd 0 = 0,
+  { have : ((cd 0 : ℂ)*z+(cd 1)).im = 0 := by simp [h],
+    simpa [z.im_ne_zero] using this },
+  ext i,
+  fin_cases i,
+  { exact this,},
+  have this1 : ((cd 0 : ℂ)*z+(cd 1)).re = 0 := by simp [h],
+  simpa [this] using this1
+end
+
+lemma bottom_ne_zero (g : SL(2, ℝ)) (z : ℍ) : bottom g z ≠ 0 :=
+linear_ne_zero (g 1) z (g.row_ne_zero 1)
+
+lemma normsq_bottom_pos (g : SL(2, ℝ)) (z : ℍ) : 0 < complex.norm_sq (bottom g z) :=
+complex.norm_sq_pos.mpr (bottom_ne_zero g z)
+
+lemma normsq_bottom_ne_zero (g : SL(2, ℝ)) (z : ℍ) : complex.norm_sq (bottom g z) ≠ 0 :=
+ne_of_gt (normsq_bottom_pos g z)
+
+/-- Fractional linear transformation -/
+def smul_aux' (g : SL(2, ℝ)) (z : ℍ) : ℂ := top g z / bottom g z
+
+lemma im_smul_eq_div_norm_sq' (g : SL(2, ℝ)) (z : ℍ) :
+  (smul_aux' g z).im = z.im / (bottom g z).norm_sq :=
+begin
+  rw [smul_aux', complex.div_im],
+  set NsqBot := (bottom g z).norm_sq,
+  have : NsqBot ≠ 0 := by simp [complex.norm_sq_pos, bottom_ne_zero g z, NsqBot, -bottom],
+  field_simp [smul_aux'],
+  convert congr_arg (λ x, x * z.im * NsqBot ^ 2) g.det_coe using 1,
+  { rw det_fin_two ↑g,
+    ring },
+  { ring }
```

Comment:
Could be
```suggestion
  rw [smul_aux', complex.div_im, ←sub_div, ←one_mul z.im, ←g.det_coe, det_fin_two, top, bottom],
  simp,
  ring,
```
or
```suggestion
  rw [smul_aux', complex.div_im, ←sub_div, ←one_mul z.im, ←g.det_coe, det_fin_two, top, bottom],
  simp only [complex.add_im, complex.mul_re, coe_fn_eq_coe, coe_im, coe_re, complex.of_real_re,
    complex.add_re, complex.of_real_im, complex.mul_im],
  ring,
```

Reply:
@Ruben-VandeVelde I often write proofs in this `convert (some multiplication) (some identity)` style to indicate that a Grobner basis tactic, when we have one, could better automate the proof.